core/src/main/scala/com/microsoft/azure/synapse/ml/explainers/ICEExplainer.scala

```diff
@@ -0,0 +1,275 @@
+        result
+          .groupBy(feature)
+          .agg(Summarizer.mean(col(targetCol)).alias(dependenceCol))
+          .agg(
+            map_from_arrays(
+              collect_list(feature),
+              collect_list(dependenceCol)
+            ).alias(outputColName)
+          )
+
+      case `individualKind` =>
+        // ICE output schema: n rows * 2 cols (idCol + ice for the given feature: map(feature_value -> explanations))
+        result
+          .groupBy(idCol)
+          .agg(
+            map_from_arrays(
+              collect_list(feature),
+              collect_list(targetCol)
+            ).alias(outputColName)
+          )
+    }
+  }
+
+  def transform(ds: Dataset[_]): DataFrame = {
+    transformSchema(ds.schema)
+    val df = ds.toDF
+    val idCol = DatasetExtensions.findUnusedColumnName("idCol", df)
+    val targetClasses = DatasetExtensions.findUnusedColumnName("targetClasses", df)
+    val dfWithId = df
+      .withColumn(idCol, monotonically_increasing_id())
+      .withColumn(targetClasses, get(targetClassesCol).map(col).getOrElse(lit(getTargetClasses)))
+
+    // Collect feature values for all features from original dataset - dfWithId
+    val (categoricalFeatures, numericFeatures) = (getCategoricalFeatures, getNumericFeatures)
+
+    // If numSamples is specified, randomly pick numSamples instances from the input dataset
+    val sampled: Dataset[Row] = get(numSamples).map(dfWithId.orderBy(rand()).limit).getOrElse(dfWithId).cache
+
+    // Collect values from the input dataframe and create dependenceDF from them
+    val calcCategoricalFunc: ICECategoricalFeature => DataFrame = {
+      f: ICECategoricalFeature =>
+        val values = collectCategoricalValues(dfWithId, f)
+        calcDependence(sampled, idCol, targetClasses, f.name, values, f.getOutputColName)
+    }
+    val calcNumericFunc: ICENumericFeature => DataFrame = {
+      f: ICENumericFeature =>
+        val values = collectSplits(dfWithId, f)
+        calcDependence(sampled, idCol, targetClasses, f.name, values, f.getOutputColName)
+    }
+
+    val dependenceDfs = (categoricalFeatures map calcCategoricalFunc) ++ (numericFeatures map calcNumericFunc)
+
+    // In the case of ICE, the function will return the initial df with columns corresponding to each feature to explain
+    // In the case of PDP the function will return df with a shape (1 row * number of features to explain)
+
+    getKind.toLowerCase match {
+      case `individualKind` =>
+        dependenceDfs.reduceOption(_.join(_, Seq(idCol), "inner"))
+          .map(sampled.join(_, Seq(idCol), "inner").drop(idCol)).get
+      case `averageKind` =>
+        dependenceDfs.reduce(_ crossJoin _)
+    }
+  }
+
+  private def collectCategoricalValues[_](df: DataFrame, feature: ICECategoricalFeature): Array[_] = {
+    val featureCountCol = DatasetExtensions.findUnusedColumnName("__feature__count__", df)
```

Comment:
nit: might want to use featureCount to be consistent with previous usage of this func